src/dag/store.rs

```diff
@@ -0,0 +1,95 @@
+use super::chunk::Chunk;
+use super::key::Key;
+use crate::kv;
+
+pub enum Error {
+    Storage(kv::StoreError),
+    CorruptStore,
+}
+
+impl From<kv::StoreError> for Error {
+    fn from(err: kv::StoreError) -> Error {
+        Error::Storage(err)
+    }
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[allow(dead_code)]
+pub struct Read {
+    kvr: Box<dyn kv::Read>,
+}
+
+#[allow(dead_code)]
+impl Read {
+    pub async fn has_chunk(&mut self, hash: &str) -> Result<bool> {
+        // Is there some way to just return the result of has() and have it automatically convert,
+        // like what happens for the error case here?
+        let has = self.kvr.has(&Key::ChunkData(hash).to_string()).await?;
+        Ok(has)
+    }
+
+    pub async fn get_chunk(&mut self, hash: &str) -> Result<Option<Chunk>> {
+        match self.kvr.get(&Key::ChunkData(hash).to_string()).await? {
+            None => Ok(None),
+            Some(data) => {
+                let meta = self.kvr.get(&Key::ChunkMeta(hash).to_string()).await?;
+                Ok(Some(Chunk::read(hash.into(), data, meta)))
+            }
+        }
+    }
+
+    pub async fn get_head(&mut self, name: &str) -> Result<Option<String>> {
+        if let Some(bytes) = self.kvr.get(&Key::Head(name).to_string()).await? {
+            match String::from_utf8(bytes) {
+                Ok(s) => return Ok(Some(s)),
+                // TODO: The detail of the error should at least get logged somewhere.
+                // Is the right thing to do to log it right here? (I guess so)
+                Err(_) => return Err(Error::CorruptStore),
+            }
+        }
+        Ok(None)
+    }
+}
+
+#[allow(dead_code)]
+pub struct Write {
+    read: Read,
+    kvw: Box<dyn kv::Write>,
+}
+
+#[allow(dead_code)]
+impl Write {
+    pub fn read(&self) -> &Read {
+        &self.read
+    }
+
+    pub async fn put_chunk(&mut self, c: Chunk) -> Result<()> {
+        self.kvw
+            .put(&Key::ChunkData(c.hash()).to_string(), c.data())
```

Comment:
I guess maybe it would be easier/better to make this key struct into a pair of bare functions or something.